File: kiko-api/src/services/judge/userSizeLayer.ts
/**
 * User Size Layer
 * Evaluates user risk profile based on trade amount
 * 
 * NOT a project risk layer - evaluates "what this amount means to the user"
 */

import {
    UserSizeLayer,
    UserSizeLevel,
    USER_SIZE_THRESHOLDS,
} from '../../types/judgeTypes.js';

/**
 * Evaluate user size layer
 * 
 * @param userAmountUsd - Trade amount in USD
 * @returns UserSizeLayer with level, score, and risk tolerance
 */
export function evaluateUserSizeLayer(userAmountUsd: number): UserSizeLayer {
    const reasons: string[] = [];

    // Determine user size level
    let level: UserSizeLevel;
    let maxSlippage: number;
    let maxRisk: number;
    let score: number;

    if (userAmountUsd < 100) {
        // L1: Small Probe - Most tolerant
        level = 'L1';
        maxSlippage = 0.15;  // 15% slippage acceptable for small tests
        maxRisk = 0.8;       // Can take high risk projects
        score = 0.9;         // Very safe for user (small amount)
        reasons.push(`Small probe ($${userAmountUsd.toFixed(2)}) - loss is manageable`);
        reasons.push('Allows higher slippage and risk tolerance');

    } else if (userAmountUsd < 400) {
        // L2: Medium - Balanced
        level = 'L2';
        maxSlippage = 0.10;  // 10%
        maxRisk = 0.6;
        score = 0.7;
        reasons.push(`Medium amount ($${userAmountUsd.toFixed(2)})`);
        reasons.push('Avoid extremely high-risk projects');

    } else if (userAmountUsd < 1500) {
        // L3: Large - More conservative
        level = 'L3';
        maxSlippage = 0.05;  // 5%
        maxRisk = 0.4;
        score = 0.5;
        reasons.push(`Large amount ($${userAmountUsd.toFixed(2)}) - be cautious`);
        reasons.push('Prefer medium/low risk projects');
        reasons.push('Slippage above 5% triggers a warning');

    } else {
        // L4: Heavy Position - Most conservative
        level = 'L4';
        maxSlippage = 0.03;  // 3%
        maxRisk = 0.25;
        score = 0.3;
        reasons.push(`Heavy position ($${userAmountUsd.toFixed(2)}) - highly cautious`);
        reasons.push('Only low-risk, high-liquidity projects are suitable');
        reasons.push('Slippage above 3% will block');
        reasons.push('Avoid very early projects (<1 hour)');
    }

    return {
        user_size_level: level,
        max_slippage_allowed: maxSlippage,
        max_risk_allowed: maxRisk,
        score,
        reasons,
    };
}

/**
 * Check if user size is compatible with given risk level
 */
export function isUserSizeCompatible(
    userSizeLayer: UserSizeLayer,
    projectRiskScore: number,  // 0-1, higher = safer
    slippageEstimate: number   // 0-1
): { compatible: boolean; warnings: string[] } {
    const warnings: string[] = [];
    let compatible = true;

    // Check slippage tolerance
    if (slippageEstimate > userSizeLayer.max_slippage_allowed) {
        warnings.push(
            `Slippage (${(slippageEstimate * 100).toFixed(1)}%) exceeds user tolerance (${(userSizeLayer.max_slippage_allowed * 100).toFixed(1)}%)`
        );
        if (userSizeLayer.user_size_level === 'L4') {
            compatible = false; // Block for heavy positions
        }
    }

    // Check risk tolerance
    // projectRiskScore is 0-1 where higher = safer
    // max_risk_allowed is 0-1 where we need (1 - projectRiskScore) < max_risk_allowed
    const projectRisk = 1 - projectRiskScore;
    if (projectRisk > userSizeLayer.max_risk_allowed) {
        warnings.push(
            `Project risk (${(projectRisk * 100).toFixed(0)}%) exceeds user tolerance (${(userSizeLayer.max_risk_allowed * 100).toFixed(0)}%)`
        );
        if (userSizeLayer.user_size_level === 'L3' || userSizeLayer.user_size_level === 'L4') {
            compatible = false;
        }
    }

    return { compatible, warnings };
}

File: kikoapi/internal/services/judge/user_size_layer.go
package judge

import (
	"fmt"

	"kikoapi/internal/types"
)

// EvaluateUserSizeLayer evaluates user risk profile based on trade amount (USD).
func EvaluateUserSizeLayer(userAmountUsd float64) types.UserSizeLayer {
	var level types.UserSizeLevel
	var maxSlippage, maxRisk, score float64
	var reasons []string

	if userAmountUsd < 100 {
		level = types.UserSizeL1
		maxSlippage = 0.15
		maxRisk = 0.8
		score = 0.9
		reasons = append(reasons, fmt.Sprintf("Small probe ($%.2f) - loss is manageable", userAmountUsd), "Allows higher slippage and risk tolerance")
	} else if userAmountUsd < 400 {
		level = types.UserSizeL2
		maxSlippage = 0.10
		maxRisk = 0.6
		score = 0.7
		reasons = append(reasons, fmt.Sprintf("Medium amount ($%.2f)", userAmountUsd), "Avoid extremely high-risk projects")
	} else if userAmountUsd < 1500 {
		level = types.UserSizeL3
		maxSlippage = 0.05
		maxRisk = 0.4
		score = 0.5
		reasons = append(reasons, fmt.Sprintf("Large amount ($%.2f) - be cautious", userAmountUsd), "Prefer medium/low risk projects", "Slippage above 5% triggers a warning")
	} else {
		level = types.UserSizeL4
		maxSlippage = 0.03
		maxRisk = 0.25
		score = 0.3
		reasons = append(reasons, fmt.Sprintf("Heavy position ($%.2f) - highly cautious", userAmountUsd), "Only low-risk, high-liquidity projects are suitable", "Slippage above 3% will block", "Avoid very early projects (<1 hour)")
	}

	return types.UserSizeLayer{
		UserSizeLevel:     level,
		MaxSlippageAllowed: maxSlippage,
		MaxRiskAllowed:    maxRisk,
		Score:             score,
		Reasons:           reasons,
	}
}

// UserSizeCompatibility is the result of checking user size vs project risk.
type UserSizeCompatibility struct {
	Compatible bool
	Warnings   []string
}

// IsUserSizeCompatible checks if user size is compatible with given risk level and slippage.
func IsUserSizeCompatible(userSizeLayer types.UserSizeLayer, projectRiskScore, slippageEstimate float64) UserSizeCompatibility {
	var warnings []string
	compatible := true

	if slippageEstimate > userSizeLayer.MaxSlippageAllowed {
		warnings = append(warnings, fmt.Sprintf("Slippage (%.1f%%) exceeds user tolerance (%.1f%%)",
			slippageEstimate*100, userSizeLayer.MaxSlippageAllowed*100))
		if userSizeLayer.UserSizeLevel == types.UserSizeL4 {
			compatible = false
		}
	}

	projectRisk := 1 - projectRiskScore
	if projectRisk > userSizeLayer.MaxRiskAllowed {
		warnings = append(warnings, fmt.Sprintf("Project risk (%.0f%%) exceeds user tolerance (%.0f%%)",
			projectRisk*100, userSizeLayer.MaxRiskAllowed*100))
		if userSizeLayer.UserSizeLevel == types.UserSizeL3 || userSizeLayer.UserSizeLevel == types.UserSizeL4 {
			compatible = false
		}
	}

	return UserSizeCompatibility{Compatible: compatible, Warnings: warnings}
}
